test/common/grpc/async_client_manager_impl_test.cc

```diff
@@ -21,6 +24,85 @@ namespace Envoy {
+  // libevent if there is a long delay between libevent regaining control. Here we want to
+  // test behavior with a specific sequence of events, where each timer fires within a
+  // simulated second of what was programmed.
+  void waitForSeconds(int seconds) {
+    for (int i = 0; i < seconds; i++) {
+      time_system_.advanceTimeAndRun(std::chrono::seconds(1), *dispatcher_,
+                                     Event::Dispatcher::RunType::NonBlock);
+    }
+  }
+
+protected:
+  Event::SimulatedTimeSystem time_system_;
+  Api::ApiPtr api_;
+  Event::DispatcherPtr dispatcher_;
+  AsyncClientManagerImpl::RawAsyncClientCache client_cache_;
+};
+
+TEST_F(RawAsyncClientCacheTest, CacheEviction) {
+  envoy::config::core::v3::GrpcService foo_service;
+  foo_service.mutable_envoy_grpc()->set_cluster_name("foo");
+  RawAsyncClientSharedPtr foo_client = std::make_shared<MockAsyncClient>();
+  client_cache_.setCache(foo_service, foo_client);
+  waitForSeconds(49);
+  // Cache entry hasn't been evicted because it was created 45s ago.
+  EXPECT_EQ(client_cache_.getCache(foo_service).get(), foo_client.get());
+  waitForSeconds(49);
+  // Cache entry hasn't been evicted because it was accessed 45s ago.
+  EXPECT_EQ(client_cache_.getCache(foo_service).get(), foo_client.get());
+  waitForSeconds(51);
+  EXPECT_EQ(client_cache_.getCache(foo_service).get(), nullptr);
+}
+
+TEST_F(RawAsyncClientCacheTest, MultipleCacheEntriesEviction) {
+  envoy::config::core::v3::GrpcService grpc_service;
+  RawAsyncClientSharedPtr foo_client = std::make_shared<MockAsyncClient>();
+  for (int i = 1; i <= 50; i++) {
+    grpc_service.mutable_envoy_grpc()->set_cluster_name(std::to_string(i));
+    client_cache_.setCache(grpc_service, foo_client);
+  }
+  waitForSeconds(20);
+  for (int i = 51; i <= 100; i++) {
+    grpc_service.mutable_envoy_grpc()->set_cluster_name(std::to_string(i));
+    client_cache_.setCache(grpc_service, foo_client);
+  }
+  waitForSeconds(30);
+  // Cache entries created 50s before have expired.
+  for (int i = 1; i <= 50; i++) {
+    grpc_service.mutable_envoy_grpc()->set_cluster_name(std::to_string(i));
+    EXPECT_EQ(client_cache_.getCache(grpc_service).get(), nullptr);
+  }
+  // Cache entries 30s before haven't expired.
+  for (int i = 51; i <= 100; i++) {
+    grpc_service.mutable_envoy_grpc()->set_cluster_name(std::to_string(i));
+    EXPECT_EQ(client_cache_.getCache(grpc_service).get(), foo_client.get());
+  }
+}
+
+TEST_F(RawAsyncClientCacheTest, GetTimeoutButNotEvictedCacheEntry) {
+  envoy::config::core::v3::GrpcService foo_service;
+  foo_service.mutable_envoy_grpc()->set_cluster_name("foo");
+  RawAsyncClientSharedPtr foo_client = std::make_shared<MockAsyncClient>();
+  client_cache_.setCache(foo_service, foo_client);
+  time_system_.advanceTimeAsyncImpl(std::chrono::seconds(50));
+  // Cache entry doesn't get evicted because it is accessed before timer fire.
```

Comment:
nit: I think the `hasn't been` phrasing used in the other test reads a bit better